backend/app/hotdog/main.py

```diff
@@ -0,0 +1,249 @@
+BUNDLE_MGR = BundleManager("/persist/bundle.json")
+
+
+def sh(cmd: str, ignore_exitcode=False):
+    p = run(split(cmd), stdout=PIPE, stderr=PIPE)
+    if not ignore_exitcode and p.returncode != 0:
+        raise Exception(p.stderr.decode())
+    return (p, p.stdout.decode(), p.stderr.decode())
+
+
+def scoped_git(cmd, scope=REPO_PATH, ignore_exitcode=False):
+    return sh(f"git -C {scope} {cmd}", ignore_exitcode)
+
+
+def is_pullable(ref):
+    return any(
+        line.endswith(ref)
+        for line in scoped_git("show-ref")[1].strip().splitlines()
+        if "refs/heads/" in line or "refs/remotes" in line
+    )
+
+
+def _corgi_checkout(ref):
+    scoped_git("fetch")
+    scoped_git(f"checkout {ref}")
+    if is_pullable(ref):
+        scoped_git("pull --rebase")
+    copytree(
+        BACKEND_REPO_DIR,
+        BACKEND_DIR,
+        dirs_exist_ok=True,
+        ignore=ignore_patterns("hotdog*"),
+    )
+    response = httpx.post(FRONTEND_HOTDOG, json={"ref": ref})
+    response.raise_for_status()
+
+
+@server.post("/checkout")
+def corgi_checkout(checkout_request: Head):
+    saved = BUNDLE_MGR.bundle.head
+    corgi_ref = saved.corgi_ref
+    enki_ref = saved.enki_ref
+    corgi_modified = saved.corgi_modified
+    enki_modified = saved.enki_modified
+    if checkout_request.corgi_ref:
+        if not REF_REGEX.match(checkout_request.corgi_ref):
+            raise Exception("Unsupported ref")
+        try:
+            _corgi_checkout(checkout_request.corgi_ref)
+            corgi_ref = checkout_request.corgi_ref
+            corgi_modified = time()
+        except Exception:
+            _corgi_checkout(saved.corgi_ref)
+            raise
+    if checkout_request.enki_ref:
+        if not REF_REGEX.match(checkout_request.enki_ref):
+            raise Exception("Unsupported ref")
+        enki_ref = checkout_request.enki_ref
+        enki_modified = time()
+    BUNDLE_MGR.save(
+        head=Head(
+            corgi_ref=corgi_ref,
+            corgi_modified=corgi_modified,
+            enki_ref=enki_ref,
+            enki_modified=enki_modified,
+        )
+    )
+
+
+@server.get("/head", response_model=Head)
+def head():
+    return BUNDLE_MGR.bundle.head
+
+
+@server.get("/")
+def home():
+    saved = BUNDLE_MGR.bundle.head
+    return HTMLResponse(
+        f"""\
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>CORGI Hotdog</title>
+</head>
+<body>
+    <h1>CORGI Hotdog</h1>
+    <form id="ref-form">
+        <label for="corgi_ref">CORGI Ref:</label>
+        <input type="text" id="corgi_ref" name="corgi_ref" placeholder="{saved.corgi_ref}">
+        <br>
+        <label for="enki_ref">Enki Ref:</label>
+        <input type="text" id="enki_ref" name="enki_ref" placeholder="{saved.enki_ref}">
+        <br>
+        <button id="submit-btn" type="submit">Submit</button>
```

Comment:
Unless there's another way to see the progress/failure of the deployment would hardcoded links to the [pipeline](https://concourse-v7.openstax.org/teams/CE/pipelines/corgi-hotdog)([s](https://concourse-v7.openstax.org/teams/CE/pipelines/corgi-hotdog-enki)) help folks know where to see the progress/failure?

Reply:
Yeah, that seems like a good idea for the time being. I have set most of the groundwork for realtime deployment status, but it will require additional work that I did not want to include at this time.

Reply:
I ended up adding the status updates, as they were initially planned, to the pipeline.